statistics/boostrap.go

```diff
@@ -27,38 +27,38 @@ import (
 	"golang.org/x/net/context"
 )
 
-func initStatsMeta4Chunk(is infoschema.InfoSchema, tables statsCache, iter *chunk.Iterator4Chunk) {
+func (h *Handle) initStatsMeta4Chunk(is infoschema.InfoSchema, tables statsCache, iter *chunk.Iterator4Chunk) {
 	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
-		tableID := row.GetInt64(1)
-		table, ok := is.TableByID(tableID)
+		physicalID := row.GetInt64(1)
+		table, ok := h.getTableByPhysicalID(is, physicalID)
 		if !ok {
-			log.Debugf("Unknown table ID %d in stats meta table, maybe it has been dropped", tableID)
+			log.Debugf("Unknown physical ID %d in stats meta table, maybe it has been dropped", physicalID)
 			continue
 		}
 		tableInfo := table.Meta()
 		newHistColl := HistColl{
-			TableID:     tableInfo.ID,
-			HaveTblID:   true,
-			Count:       row.GetInt64(3),
-			ModifyCount: row.GetInt64(2),
-			Columns:     make(map[int64]*Column, len(tableInfo.Columns)),
-			Indices:     make(map[int64]*Index, len(tableInfo.Indices)),
-			colName2Idx: make(map[string]int64, len(tableInfo.Columns)),
-			colName2ID:  make(map[string]int64, len(tableInfo.Columns)),
+			PhysicalID:     physicalID,
+			HavePhysicalID: true,
+			Count:          row.GetInt64(3),
+			ModifyCount:    row.GetInt64(2),
+			Columns:        make(map[int64]*Column, len(tableInfo.Columns)),
+			Indices:        make(map[int64]*Index, len(tableInfo.Indices)),
+			colName2Idx:    make(map[string]int64, len(tableInfo.Columns)),
+			colName2ID:     make(map[string]int64, len(tableInfo.Columns)),
 		}
 		tbl := &Table{
 			HistColl: newHistColl,
 			Version:  row.GetUint64(0),
 		}
-		tables[tableID] = tbl
+		tables[physicalID] = tbl
 	}
 }
 
 func (h *Handle) initStatsMeta(is infoschema.InfoSchema) (statsCache, error) {
 	h.mu.Lock()
-	defer h.mu.Unlock()
 	sql := "select version, table_id, modify_count, count from mysql.stats_meta"
 	rc, err := h.mu.ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), sql)
+	h.mu.Unlock()
```

Comment:
Why unlock earlier can fix data race?

Reply:
The data race is fixed by moving the `pid2tid` to `mu` in [here](https://github.com/pingcap/tidb/pull/7305/files#diff-465d91625076a342f8a71042886eeae1R45). But after that, the `initStatsMeta4Chunk` will call [getTableByPhysicalId](https://github.com/pingcap/tidb/pull/7305/files#diff-b34e30469b13b5f31b540b13594a7a59R33), so if we still hold the lock, it will cause dead lock.